crates/sparrow-runtime/src/prepare.rs

```diff
@@ -140,192 +137,92 @@ impl<'a> fmt::Display for SourceDataWrapper<'a> {
-    let (pm, pf): (Vec<_>, Vec<_>) = results.try_collect::<Vec<_>>().await?.into_iter().unzip();
-    match output_key {
-        ObjectStoreKey::Local | ObjectStoreKey::Memory => {
-            // Prepared files are stored locally
-            Ok((pm, pf))
-        }
-        _ => {
-            // Upload the local prepared files to the remote destination
-            let upload_results = upload_prepared_files(
-                object_store_registry,
-                &pm,
-                output_path_prefix,
-                output_file_prefix,
-            )
-            .await?;
-            Ok((pm, upload_results))
-        }
-    }
-}
-
-async fn upload_prepared_files(
-    object_store_registry: &ObjectStoreRegistry,
-    prepared_metadata: &[PreparedMetadata],
-    output_path_prefix: &str,
-    output_file_prefix: &str,
-) -> error_stack::Result<Vec<PreparedFile>, Error> {
-    let span = tracing::info_span!(
-        "Uploading prepared files to object store",
-        ?output_path_prefix
-    );
-    let _enter = span.enter();
-
-    let mut parquet_uploads: Vec<_> = Vec::new();
-    let mut prepared_files: Vec<_> = Vec::new();
-    {
-        for (batch_count, prepared_metadata) in prepared_metadata.iter().enumerate() {
-            tracing::debug!(
-                "Prepared batch {batch_count} has {} rows",
-                prepared_metadata.num_rows
-            );
-
-            tracing::debug!(
-                "Prepared batch {batch_count} with metadata {:?}",
-                prepared_metadata
-            );
-
-            let prepared_url =
-                format!("{output_path_prefix}/{output_file_prefix}-{batch_count}.parquet");
-            let metadata_prepared_url =
-                format!("{output_path_prefix}/{output_file_prefix}-{batch_count}-metadata.parquet");
-
-            let prepare_object_store_url = ObjectStoreUrl::from_str(prepared_url.as_str())
-                .change_context_lazy(|| Error::InvalidUrl(prepared_url.as_str().to_string()))?;
-            let metadata_object_store_url =
-                ObjectStoreUrl::from_str(metadata_prepared_url.as_str()).change_context_lazy(
-                    || Error::InvalidUrl(metadata_prepared_url.as_str().to_string()),
-                )?;
-
-            parquet_uploads.push(
-                object_store_registry
-                    .upload(prepare_object_store_url, Path::new(&prepared_metadata.path)),
-            );
-
-            parquet_uploads.push(object_store_registry.upload(
-                metadata_object_store_url,
-                Path::new(&prepared_metadata.metadata_path),
-            ));
-
-            let result_prepared_file: PreparedFile = prepared_metadata
-                .to_owned()
-                .with_path(prepared_url)
-                .with_metadata_path(metadata_prepared_url)
+        // Create the prepared file via PreparedMetadata.
+        // TODO: We could probably do this directly, eliminating the PreparedMetadata struct.
```

Comment:
What is the work required to eliminate the prepared metadata struct? It's odd to go from PreparedMetadata back to the PreparedFile.

Reply:
We were always doing that -- in fact, for a while we were returning both and dropping one of them, so it was just "hidden". I think it could be a few parts:

1. We could just add some methods to the `PreparedFile` (the proto equivalent) of `PreparedMetadata`, which AFAICT isn't widely used outside of "get the schema of this ParquetFile" which we can do better using the `ParquetFile` struct.
2. We could also go even further -- technically, we use the prepared file to store the min/max time and the num rows. That is available directly in the ParquetMetadata, which in turn we can get from the "footer" of the file. Currently, we need to fetch the footer to initialize the `ParquetFile` (and determine the schema), but we could put that footer in the DB, which would allow us to skip that step entirely.

I don't think this should be done in this PR though. Thoughts?

Reply:
Agreed, I was just curious to see the effort. Its also not worth prioritizing given the limited usage of the prepared metadata. Looks good to me!